torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <unordered_set>
+#include <vector>
+
+#include "torch/csrc/utils/functional.h"
+
+namespace torch { namespace jit { namespace detail {
+
+// DynamicDAG is a DAG that dynamically maintains a topological order as
+// edges/vertices are added and removed.
+//
+// The implementation is based off of the following paper:
+// "A Dynamic Topsort Algorithm for Directed Acyclic Graphs"
+// by David Pearce and Paul Kelly
+// https://www.doc.ic.ac.uk/~phjk/Publications/DynamicTopoSortAlg-JEA-07.pdf
+
+template <typename T> struct Vertex;
+
+template <typename T>
+using vertex_set = std::unordered_set<Vertex<T>*>;
+
+template <typename T>
+using vertex_list = std::vector<Vertex<T>*>;
+
+template <typename T>
+using maybe_vertex_list = std::vector<at::optional<Vertex<T>*>>;
+
+template <typename T>
+struct Vertex {
+  std::vector<T> rdata; // stored in reverse order
```

Comment:
In reverse order wrt... what order?

Reply:
Reverse order to the order in which the Node* appear in the Graph&.